src/coprocessor/codec/convert.rs

```diff
@@ -45,51 +86,157 @@ pub fn truncate_f64(mut f: f64, flen: u8, decimal: u8) -> Res<f64> {
 
 /// `overflow` returns an overflowed error.
 macro_rules! overflow {
-    ($val:ident, $bound:ident) => {{
-        Err(box_err!("constant {} overflows {}", $val, $bound))
+    ($val:ident, $bound:tt) => {{
+        Error::Eval(
+            format!("constant {} overflows {}", $val, $bound),
+            ERR_DATA_OUT_OF_RANGE,
+        )
     }};
 }
 
+/// `convert_int_to_int` converts an int value to a diferent int value.
+#[inline]
+pub fn convert_int_to_int(ctx: &mut EvalContext, val: i64, tp: FieldTypeTp) -> Result<i64> {
+    let lower_bound = integer_signed_lower_bound(tp);
+    // https://dev.mysql.com/doc/refman/8.0/en/out-of-range-and-overflow.html
+    if val < lower_bound {
+        ctx.handle_overflow(overflow!(val, lower_bound))?;
+        return Ok(lower_bound);
+    }
+    let upper_bound = integer_signed_upper_bound(tp);
+    if val > upper_bound {
+        ctx.handle_overflow(overflow!(val, upper_bound))?;
+        return Ok(upper_bound);
+    }
+    Ok(val)
+}
+
 /// `convert_uint_to_int` converts an uint value to an int value.
 #[inline]
-pub fn convert_uint_to_int(val: u64, upper_bound: i64, tp: FieldTypeTp) -> Result<i64> {
+pub fn convert_uint_to_int(ctx: &mut EvalContext, val: u64, tp: FieldTypeTp) -> Result<i64> {
+    let upper_bound = integer_signed_upper_bound(tp);
     if val > upper_bound as u64 {
-        return overflow!(val, tp);
+        ctx.handle_overflow(overflow!(val, upper_bound))?;
+        return Ok(upper_bound);
     }
     Ok(val as i64)
 }
 
 /// `convert_float_to_int` converts an f64 value to an i64 value.
 ///  Returns the overflow error if the value exceeds the boundary.
-pub fn convert_float_to_int(
-    fval: f64,
-    lower_bound: i64,
-    upper_bound: i64,
-    tp: FieldTypeTp,
-) -> Result<i64> {
+#[inline]
+pub fn convert_float_to_int(ctx: &mut EvalContext, fval: f64, tp: FieldTypeTp) -> Result<i64> {
     // TODO any performance problem to use round directly?
     let val = fval.round();
```

Comment:
Is this round behaviour the same as RoundFloat in TiDB?

Reply:
Seems no difference.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f7bed4f47e4b30a2a1d15bb6cf73532e
```rust
use std::f64;

fn main() {
    let tests: Vec<f64> = vec![0.1, 1.4,1.5,1.6, -0.1, -0.5, -0.6, -1.5, -1.6];
    for x in tests {
        println!("{} -> {}", x, x.round());
    }
}
```
output:
```
0.1 -> 0
1.4 -> 1
1.5 -> 2
1.6 -> 2
-0.1 -> 0
-0.5 -> -1
-0.6 -> -1
-1.5 -> -2
-1.6 -> -2
```

https://play.golang.org/p/ewxOmdFPb7D
```go
package main

import (
	"fmt"
	"math"
)

func RoundFloat(f float64) float64 {
	if math.Abs(f) < 0.5 {
		return 0
	}

	return math.Trunc(f + math.Copysign(0.5, f))
}

func main() {
	var tests = []float64 {
		0.1, 1.4,1.5,1.6, -0.1, -0.5, -0.6, -1.5, -1.6,
	}
	for _, x := range tests {
		fmt.Printf("%v -> %v\n", x, RoundFloat(x))
	}
}
```
```
0.1 -> 0
1.4 -> 1
1.5 -> 2
1.6 -> 2
-0.1 -> 0
-0.5 -> -1
-0.6 -> -1
-1.5 -> -2
-1.6 -> -2
```